util/io/paths.go

```diff
@@ -0,0 +1,37 @@
+package io
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+// TempPaths allows generating and memorizing random paths for a given URL.
+type TempPaths struct {
+	paths map[string]string
+	lock  sync.Mutex
+}
+
+func NewTempPaths() *TempPaths {
+	return &TempPaths{
+		paths: make(map[string]string),
+	}
+}
+
+// GetPath generates a path for the given URL or returns previously generated one.
+func (p *TempPaths) GetPath(url string) (string, error) {
```

Comment:
Maybe `url` parameter should be named to something more universal, such as `key`